ross/defects/integrate_solver.py

```diff
@@ -0,0 +1,142 @@
+"""Time integration module.
+
+This module defines the time integration routines to evaluate time responses from
+the rotors. These simulate the transient evolution of the dynamic behavior of the
+rotors defined by the user, and return the time signals for rotor displacement.
+
+"""
+
+import numpy as np
+
+import ross
+
+
+class Integrator:
+
+    """A series of Runge-Kutta time integration algorithms.
+    
+    Calculates the time response for the rotors input to the routine.
+
+    Parameters
+    ----------
+    x0 : float
+        Initial time
+    y0 : float
+        Initial condition for the integration
+    x : float
+        Iteration time
+    h : float
+        Step height
+    func : object
+        Function to be integrated in time
+
+    Returns
+    -------
+    The rotor transient response.
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Runge–Kutta_methods ..
+
+    """
+
+    def __init__(
+        self, x0, y0, x, h, func,
+    ):
+
+        self.x0 = x0
+        self.y0 = y0
+        self.x = x
+        self.h = h
+        self.func = func
+
+    def rk4(self):
+        # Runge-Kutta 4th order (RK4)
+
+        # Count number of iterations using step size or
+        # step height h
+        n = int((self.x - self.x0) / self.h)
+
+        # Iterate for number of iterations
+        y = self.y0
+        result = np.zeros((24, n + 1))
+        result[:, 0] = self.y0
+
+        # 4th-order Runge-Kutta
+
+        for i in range(1, n + 1):
+            if i % 10000 == 0:
+                print(f"Iteration: {i} \n Time: {self.x0}")
+
+            "Apply Runge Kutta Formulas to find next value of y"
+            k1 = self.h * self.func(self.x0, y, i)
+            k2 = self.h * self.func(self.x0 + 0.5 * self.h, y + 0.5 * k1, i)
+            k3 = self.h * self.func(self.x0 + 0.5 * self.h, y + 0.5 * k2, i)
+            k4 = self.h * self.func(self.x0 + self.h, y + k3, i)
+
+            # Update next value of y
+            y = y + (1.0 / 6.0) * (k1 + 2 * k2 + 2 * k3 + k4)
+            result[:, i] = np.copy(y)
+
+            # Update next value of x
+            self.x0 = self.x0 + self.h
+
+        return result
+
+    def rk45(self):
```

Comment:
This is just a comment for our evaluation in the future.
From what I understand, although this integrator is available in the scipy package, you had to implement it here to have access to some internal calculation. I was searching for a way to do this with the scipy function and one possibility was to use global variables. I have tested nothing, but this could be an option, so we could use the scipy code without the extra work of testing/maintaining the integrator.

Reply:
We attempted to solve the problem by at least 7 different types of integrators, and despite trying the RK45 from scipy with all the options we could it simply did not converge. The most robust and reliable method was our own RK45, with constant time step.